safehttp/incoming_request.go

```diff
@@ -14,9 +14,80 @@
 
 package safehttp
 
-import "net/http"
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+)
 
 // IncomingRequest TODO
 type IncomingRequest struct {
-	req *http.Request
+	req       *http.Request
+	parseOnce sync.Once
+}
+
+// QueryForm parses the query parameters provided in the request. It returns
+// the parsed parameters as a Form object, if no error occured, or the parsing error otherwise.
+func (r *IncomingRequest) QueryForm() (f *Form, err error) {
+	r.parseOnce.Do(func() {
+		if r.req.Method != "GET" {
+			f, err = nil, fmt.Errorf("got request method %s, want GET", r.req.Method)
+		}
+		if err := r.req.ParseForm(); err != nil {
+			f = nil
+		}
+	})
+	f, err = &Form{values: r.req.Form}, nil
+	return
+}
+
+// PostForm parses the form parameters provided in the body of a POST, PATCH or
+// PUT request that does not have Content-Type: multipart/form-data. It returns
+// the parsed parameters as a Form object, if no error occured, or the parsing error otherwise.
+func (r *IncomingRequest) PostForm() (f *Form, err error) {
+	r.parseOnce.Do(func() {
+		if r.req.Method != "POST" && r.req.Method != "PATCH" && r.req.Method != "PUT" {
+			f, err = nil, fmt.Errorf("got request method %s, want POST/PATCH/PUT", r.req.Method)
+		}
+
+		if ct := r.req.Header.Get("Content-Type"); ct != "application/x-www-form-urlencoded" {
+			f, err = nil, fmt.Errorf("invalid method called for Content-Type: %s, want MultipartForm", ct)
+		}
+		if err := r.req.ParseForm(); err != nil {
+			f = &Form{}
+		}
+	})
+	f, err = &Form{values: r.req.PostForm}, nil
+	return
+}
+
+// MultipartForm parses the form parameters provided in the body of a POST,
+// PATCH or PUT request that has Content-Type set tomultipart/form-data. It
+// returns a MultipartForm object containing the parsed form parameter and
+// files, if no error occured, or the parsing error otherwise.
+func (r *IncomingRequest) MultipartForm(maxMemory int64) (f *MultipartForm, err error) {
+	r.parseOnce.Do(func() {
+		const defaultMaxMemory = 32 << 20
```

Comment:
`32 << 20` seems like a pretty arbitrary limit. 33 megabytes. Is there any reason behind picking this number?
It should probably be mentioned in the comment above that `maxMemory` is capped at this value.

Reply:
That is the maxDefaultMemory in the `net/http` implementation so it's mostly to be backwards compatible, but I guess it would be a good thing  to look more into it.

Reply:
Yep, indeed, resolved in [0f32eb1](url)

Reply:
This is not resolved in 0f32eb1. The godoc comment still doesn't mention the capping of `maxMemory`

Reply:
Thanks, sorry! :) [b3cb723](url) resolves it

Reply:
I don't get this. Why is this capped to 32MB? Just because it's the default, doesn't mean it should be a hard cap.

Reply:
I'm going to merge this PR but open an issue for this.,